### PR TITLE
[Review] Cyborg headlamps!

### PR DESCRIPTION
**Pull request by GunHog**
Cyborgs now have built in headlamps, controlled by a screen button. 
~~Lighting intensity for cyborgs are the same values as they were previously, including borgs that had different flashlight types.~~

Sprites by @xxalpha!

The headlamp has five power levels, most having a power cost:
- Level 1, half a standard flashlight, costing zero power and showing no light sprite.
- Level 2, standard flashlights, 4W per tick power cost. Sprite shows lights.
- Level 3, mining lantern level, 8W per tick.
- Level 4, high power level, 12W per tick.
- Level 5, floodlights, 16W per tick. Use sparingly!

Being put into a stat (stuns) will disable the lamp, including being hit with EMP and having no cell.

Cyborgs no longer have a flashlight module, as they are no longer needed!
- Partially addresses https://github.com/tgstation/-tg-station/issues/10288.

Borgies!! (Ignore the chat, that is debug text only!)
screenshot 2015-07-14 10 04 49

Service Borgies!
screenshot 2015-07-14 16 06 14

### CODE REVIEW

**Comment by kevinz000** [comment]
secborgs should have a brighter lamp
Because they had a seclite
Seclites are brighter then flashlights

**Comment by GunHog** [comment]
Mining borgs are set to 6 (lantern).
Security borgs are set to 5 (seclite).
All others are set to 4 (standard).

**Comment by Metacide** [comment]
Great, looks much more convenient :+1:

**Comment by Supermichael777** [comment]
halajuja

**Comment by xxalpha** [comment]
:+1:

**Comment by Cebasiz** [comment]
good change, juggling flashlights was always a pain

**Comment by GunHog** [comment]
If one of you posts a button sprite, I will replace it. I could go with a neat little sound too.

**Comment by Niknakflak** [comment]
Should be called Eyelamps because glowey eyes would be cool.  Bonus if you actually make the eyes light up and look off when not in use.  Headlamps just doesn't seem as cool.  Although this may be a problem for cyborgs with smaller eye sprites.

**Comment by GunHog** [comment]
Weeeeell...if I had....sprites....

**Comment by freerealestate** [comment]
:+1: Great work

**Comment by Fox-McCloud** [comment]
da powercreep

**Comment by freerealestate** [comment]
Being able to see where you're going shouldn't be considered power, it should be considered basic functionality.

**Comment by optimumtact** [comment]
It should take power though

**Comment by GunHog** [comment]
Current flashlights do not take power. I am not against this though, I originally coded it that way, but if I put it back, I will have to do something about it being possible to never completely drain the cell if I use a number bigger than 1.

**Comment by optimumtact** [comment]
oh I had no idea

**Comment by xxalpha** [comment]
What if
guys

What if cyborgs could control their light range, power consumption increases with range?

**Comment by GunHog** [comment]
I will have to look into that.

**Comment by Dorsisdwarf** [comment]
Seriously though I love the idea of making the lamps be glowing eyes

**Comment by GunHog** [comment]
If I had sprites for each borg...

**Comment by GunHog** [comment]
Although, I do not want to make this overtly complicated. Keeping it simple as I have it now is also an option. Perhaps I should make a forum thread and get opinions? I am happy to take sprites for the button and bright borg eyes, though!

**Comment by optimumtact** [comment]
Uh I pretty much main silicon

**Comment by GunHog** [comment]
Just letting everyone know my progress:
- The system now has select-able light levels from 2 to 10
- Pretty buttons by xxalpha which change to reflect the current setting!
- Sprites for the borgies themselves!

I need to finish the power drain system, as well as define conditions when the lamp is forcibly deactivated (not enough power, unconscious, etc).

**Comment by GunHog** [comment]
<b>This is only a progress update, not the final product.</b>
I switched to a power-draining system, with the weakest setting being free. The strongest will GORGE itself on the borg's cell, and is especially taxing on Security and Engineering borgs, who have the costliest modules and functions.

xxalpha provided some lovely sprites, so I have added them here as well! The lights show up at the Level 2 setting or higher.

There are still some things I must fix and tweak before a proper review:
[ ] Fix the light button not updating when disabled forcibly.
[ ] Fix the lack of light overlays for borgs that recover from a stun.
[ ] Improve the power drain code
[ ] Ensure that the Shadowling's Veil power is supported.
[ ] Perhaps have cooldown based code preventing light use after being disabled.

**Comment by Dorsisdwarf** [comment]
This is really nice

**Comment by GunHog** [comment]
I now need a maintainer review for what ways to improve my code and bring it up to standard, plus any other fixes/suggestions.

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 1]
```diff
@@ -978,6 +987,36 @@
...
 
+/mob/living/silicon/robot/proc/control_headlamp()
+	if(stat)
+		src << "<span class='danger'>This function is currently offline.</span>"
+		return
+
+	if(lamp_intensity < lamp_max) //Lamp is set lower than the max.
+		lamp_intensity += 2
+		src << "Headlamp power set to Level [lamp_intensity/2]."
+	else //Greater than or equal to the max, so turn it off instead.
```
Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: you can use modulo for this, as this is clearly clock arithmetic.

lamp_intensity = lamp_intensity % your_max + 1
This works like this:
1 = 1 % 5 +1 = 2
2 = 2 % 5 +1 = 3
3 = 3 % 5 +1 = 4
4 = 4 % 5 + 1 = 5
5 = 5 % 5 + 1 = 1 (X Mod X is always 0)

this is a MUCH nicer way to format a "cycle" between 1 and a Max value.

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 2]
```diff
@@ -978,6 +987,36 @@
...
+	update_headlamp()
+
+/mob/living/silicon/robot/proc/update_headlamp(var/turn_off = 0)
+	SetLuminosity(0)
+
+	if(lamp_intensity && (turn_off || stat))
+		src << "<span class='danger'>Your headlamp has been deactivated.</span>"
+		lamp_intensity = 0
+		if(client)
+			var/obj/screen/robot/lamp/button = locate(/obj/screen/robot/lamp) in client.screen
```
Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: locate is slower than a for loop.
eg:
```
var/the_button = null
for(var/obj/screen/robot/lamp/button in client.screen)
    the_button = button
    break
```
is faster than ```locate(/obj/screen/robot/lamp) in client.screen```

Comment (GunHog) [inline code/modules/mob/living/silicon/robot/robot.dm]: I have seen other maintainers say not to use for() loops when hunting for only one element in something. Has this precedent changed?

Comment (xxalpha) [inline code/modules/mob/living/silicon/robot/robot.dm]: Can't you have an association?

Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: The precedent has not changed, but it should, finding one element in a list is presently faster with a for()+break than with locate(). if you worry about this, then go with option B, a permanent reference.

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 1 above)
Comment (GunHog) [inline code/modules/mob/living/silicon/robot/robot.dm]: I need this to work in increments of 2.

Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: your max is 10, thus even, meaning you can swap the +1 for +2, duh.

**Comment by GunHog** [comment]
Alrighty, RR's magical modulo thingummywut that took me 10 bloody minutes to make work in increments of two (I am <b>AWFUL</b> at math!) is implemented. Also, I associated by lamp and button directly, so that works better now as well!

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 3]
```diff
@@ -978,6 +988,32 @@
 	set category = "Robot Commands"
 	set_autosay()
 
+/mob/living/silicon/robot/proc/control_headlamp()
+	if(stat)
+		src << "<span class='danger'>This function is currently offline.</span>"
+		return
+
+//Some sort of magical "modulo" thing which somehow increments lamp power by 2, until it hits the max and resets to 0.
+	lamp_intensity = (lamp_intensity+2) % (lamp_max+2)
```
Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: ```lamp_intensity = lamp_intensity % lamp_max + 2``` works

Comment (GunHog) [inline code/modules/mob/living/silicon/robot/robot.dm]: That does not work. It returns 1 when it reaches the max instead of 0.

Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: the max IS 10 yeah? 10 mod 10 (lamp_intensity % lamp_max) is 0, not 1

Comment (GunHog) [inline code/modules/mob/living/silicon/robot/robot.dm]: I meant to say 2, not 1. It will do the 10 mod-whaever 10, then add 2 at the end, giving 2.

Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: yeah yeah it's k, this code is fine now, you've used the modulo operator to clean it up and it works.

Comment (GunHog) [inline code/modules/mob/living/silicon/robot/robot.dm]: Please do not do that to me again, RR. Math is evil and scary. I would rather re-write Changeling.

Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: rofl.

Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: and if you didn't know, for learning purposes:

```
0+2 % 12 = 2
2+2 % 12 = 4
4+2 % 12 = 6
6+2 % 12 = 8
8+2 % 12 = 10
10+2 % 12 = 0
```
is how your current code runs, which is what you wanted.
x mod y is x so long as x is not y, or 
```
var/result = 0
if(x != y) 
    result = x
else
    result = 0
```

**Comment by borntohonk** [comment]
I like it.

**Pull request merged**